File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
class Solution(object):
    def rec(self, root, i):
        if root is None:
            return i
            
        i += 1
        left_depth = self.rec(root.left, i)
        right_depth = self.rec(root.right, i)
        return max(left_depth, right_depth)

    def maxDepth(self, root):
        if root is None:
            return 0
        return self.rec(root, 0)

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (!root) return 0;

        int depth = 1, maxDepth = depth;
        check(root, depth, maxDepth);
        return maxDepth;
    }

    void check(TreeNode* root, int depth, int& maxDepth){
        if (!root) return;

        check(root->left, depth + 1, maxDepth);
        maxDepth = max(depth, maxDepth);
        check(root->right, depth + 1, maxDepth);
    }
};